### PR TITLE
Bugfix: prevent creation of multiple search fields after cancel opera…

### PR DESCRIPTION
**Pull request by scholtzan**
Fixes https://github.com/xi-editor/xi-mac/issues/349

## Review Checklist

- [ ] I have responded to reviews and made changes where appropriate.
- [x] I have tested the code
- [x] I have updated comments / documentation related to the changes I made.
- [x] I have rebased my PR branch onto xi-mac/master.

### CODE REVIEW

File: Sources/XiEditor/EditViewController.swift [hunk 1]
```diff
@@ -442,6 +443,7 @@ class EditViewController: NSViewController, EditViewDataSource, FindDelegate, Sc
         if !findViewController.view.isHidden {
             closeFind()
         } else {
+            unsetFind()
```
[the remark this answers is not in the document]
Comment (scholtzan) [reply Sources/XiEditor/EditViewController.swift]: Actually, thinking about this, `cancel_operation` can get removed altogether (updated this PR, can open a PR for removing it in core if this gets merged). 

The problem was that `cancel_operation` deleted all `Find`s in core but xi-mac did not delete them. So for the next find action a new `Find` was created with a new ID and then added as additional search field in xi-mac.

[the remark this answers is not in the document]
Comment (scholtzan) [reply Sources/XiEditor/EditViewController.swift]: Yup. In a way `unsetFind` is basically doing the same thing as `cancel_operation`. It removes all search fields and makes sure core does the same by sending `find`.

[the remark this answers is not in the document]
Comment (cmyr) [reply Sources/XiEditor/EditViewController.swift]: okay playing around with this, the first thing I notice is that if I just select a word and press esc we no longer collapse selections. I think we do want this behaviour?

[the remark this answers is not in the document]
Comment (cmyr) [reply Sources/XiEditor/EditViewController.swift]: and maybe worse, <kbd>esc</kbd> now removes focus from the window.

[the remark this answers is not in the document]
Comment (scholtzan) [reply Sources/XiEditor/EditViewController.swift]: Oh right, so I guess `cancel_operation` has to stay for now

[the remark this answers is not in the document]
Comment (cmyr) [reply Sources/XiEditor/EditViewController.swift]: <kbd>esc</kbd> behaviour is tricky, and it's very possible we shouldn't be sending it as a general purpose command and should be deciding exactly what we mean in a given context; but removing it right now is a usability regression I think 😞

[the remark this answers is not in the document]
Comment (scholtzan) [reply Sources/XiEditor/EditViewController.swift]: Yeah 😞 back to using `cancel_operation`

[the remark this answers is not in the document]
Comment (cmyr) [reply Sources/XiEditor/EditViewController.swift]: On this last commit I think I'm still seeing the thing where if I press esc I lose focus? with or without active selections..

**Comment by scholtzan** [comment]
Okay, I think I came up with a proper fix for this. Now, based on the find status all finds that have previously been deleted in the core will also get removed in the frontend.
For this to work https://github.com/xi-editor/xi-editor/pull/1013 is required (after `unset_find` needs to send the new find status).

**Comment by cmyr** [comment]
Okay, sounds good. I think that my main issue is a secondary bug; <kbd>cmd</kbd><kbd>D</kbd> with the find view not visible is still changing focus to the find view, maybe? I think this is why <kbd>esc</kbd> isn't working as I expect.

File: Sources/XiEditor/Search.swift [hunk 2]
```diff
@@ -321,6 +321,10 @@ extension EditViewController {
             updateScrollPosition(previousOffset: findViewController.view.fittingSize.height)
         }
 
+        if findViewController.searchQueries.isEmpty {
+            findViewController.addSearchField(searchField: nil)
```
Comment (cmyr) [inline Sources/XiEditor/Search.swift]: okay, there's a bug here: `addSearchField` changes the first responder, even if the find view isn't currently visible. This means that if cmd+D happens when  there are no queries, the editor window loses focus and you have to manually click on the view to regain.

I think this points at some architectural problems; the focus change really shouldn't be a side effect of a function that might be called in different places; so maybe either `addSearchField` needs to be split up, or the focus change should be moved to somewhere else?

Comment (scholtzan) [reply Sources/XiEditor/Search.swift]: Maybe simply an additional `makeFirstResponder` parameter is sufficient for `addSearchField`?

Comment (cmyr) [reply Sources/XiEditor/Search.swift]: You could also do a check and only call `makeFirstResponder` if the view controller is visible.

Long-term there's some code smell here related to how we track state; the current control flow is pretty tough to follow, and I think that's the underlying cause of our original bug.

**Pull request merged**